File: radar-commons-android/src/main/java/org/radarcns/android/data/TableDataHandler.java
/*
 * Copyright 2017 The Hyve
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.radarcns.android.data;

import android.content.Context;
import android.content.Intent;
import android.os.Handler;
import android.os.HandlerThread;
import android.support.annotation.NonNull;
import android.util.Pair;

import org.apache.avro.specific.SpecificData;
import org.apache.avro.specific.SpecificRecord;
import org.radarcns.android.auth.AppAuthState;
import org.radarcns.android.kafka.KafkaDataSubmitter;
import org.radarcns.android.kafka.ServerStatusListener;
import org.radarcns.android.util.AtomicFloat;
import org.radarcns.android.util.BatteryLevelReceiver;
import org.radarcns.android.util.NetworkConnectedReceiver;
import org.radarcns.config.ServerConfig;
import org.radarcns.kafka.ObservationKey;
import org.radarcns.producer.rest.RestClient;
import org.radarcns.producer.rest.RestSender;
import org.radarcns.producer.rest.SchemaRetriever;
import org.radarcns.topic.AvroTopic;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.IOException;
import java.util.ArrayList;
import java.util.Collections;
import java.util.HashSet;
import java.util.List;
import java.util.Locale;
import java.util.Map;
import java.util.Set;
import java.util.TreeMap;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicBoolean;
import java.util.concurrent.atomic.AtomicLong;

import static org.radarcns.android.device.DeviceService.CACHE_RECORDS_SENT_NUMBER;
import static org.radarcns.android.device.DeviceService.CACHE_RECORDS_UNSENT_NUMBER;
import static org.radarcns.android.device.DeviceService.CACHE_TOPIC;

/**
 * Stores data in databases and sends it to the server.
 */
public class TableDataHandler implements DataHandler<ObservationKey, SpecificRecord>, BatteryLevelReceiver.BatteryLevelListener, NetworkConnectedReceiver.NetworkConnectedListener {
    private static final Logger logger = LoggerFactory.getLogger(TableDataHandler.class);

    private static final long DATA_RETENTION_DEFAULT = 86400000L;
    private static final int SEND_LIMIT_DEFAULT = 1000;
    private static final long UPLOAD_RATE_DEFAULT = 10L;
    private static final long SENDER_CONNECTION_TIMEOUT_DEFAULT = 10L;
    private static final float MINIMUM_BATTERY_LEVEL = 0.1f;
    private static final float REDUCED_BATTERY_LEVEL = 0.2f;

    private final Map<String, DataCacheGroup<ObservationKey, ? extends SpecificRecord>> tables = new ConcurrentHashMap<>();
    private ServerStatusListener statusListener;
    private final Object STATUS_SYNC = new Object();
    private final BatteryLevelReceiver batteryLevelReceiver;
    private final NetworkConnectedReceiver networkConnectedReceiver;
    private final AtomicBoolean sendOnlyWithWifi;
    private final Context context;
    private final SpecificData specificData;
    private Handler handler;
    private final HandlerThread handlerThread;
    private int maxBytes;
    private AppAuthState authState;
    private ServerConfig kafkaConfig;
    private SchemaRetriever schemaRetriever;
    private boolean hasBinaryContent;
    private final AtomicBoolean sendOverDataHighPriority;
    private final Set<String> highPriorityTopics;
    private int kafkaRecordsSendLimit;
    private final AtomicLong dataRetention;
    private long kafkaUploadRate;
    private long senderConnectionTimeout;

    private ServerStatusListener.Status status;
    private final Map<String, Integer> lastNumberOfRecordsSent = new TreeMap<>();
    private KafkaDataSubmitter submitter;
    private RestSender sender;
    private final AtomicFloat minimumBatteryLevel;
    private boolean useCompression;

    /**
     * Create a data handler. If kafkaConfig is null, data will only be stored to disk, not uploaded.
     */
    public TableDataHandler(Context context, ServerConfig kafkaUrl, SchemaRetriever schemaRetriever,
                            int maxBytes, boolean sendOnlyWithWifi, boolean hasBinaryContent,
                            boolean sendOverDataHighPriority, Set<String> highPriorityTopics,
                            AppAuthState authState) {
        this.context =  context;
        this.kafkaConfig = kafkaUrl;
        this.schemaRetriever = schemaRetriever;
        this.hasBinaryContent = hasBinaryContent;
        this.sendOverDataHighPriority = new AtomicBoolean(sendOverDataHighPriority);
        this.highPriorityTopics = new HashSet<>(highPriorityTopics);
        this.kafkaUploadRate = UPLOAD_RATE_DEFAULT;
        this.kafkaRecordsSendLimit = SEND_LIMIT_DEFAULT;
        this.senderConnectionTimeout = SENDER_CONNECTION_TIMEOUT_DEFAULT;
        this.minimumBatteryLevel = new AtomicFloat(MINIMUM_BATTERY_LEVEL);
        this.handlerThread = new HandlerThread("TableDataHandler");
        this.handlerThread.start();
        this.handler = new Handler(handlerThread.getLooper());
        this.handler.postDelayed(new Runnable() {
            @Override
            public void run() {
                for (ReadableDataCache cache : getCaches()) {
                    Pair<Long, Long> records = cache.numberOfRecords();
                    Intent numberCached = new Intent(CACHE_TOPIC);
                    numberCached.putExtra(CACHE_TOPIC, cache.getReadTopic().getName());
                    numberCached.putExtra(CACHE_RECORDS_UNSENT_NUMBER, records.first);
                    numberCached.putExtra(CACHE_RECORDS_SENT_NUMBER, records.second);
                    context.sendBroadcast(numberCached);
                }
                synchronized (TableDataHandler.this) {
                    if (handler != null) {
                        handler.postDelayed(this, 10_000);
                    }
                }
            }
        }, 10_000L);

        this.batteryLevelReceiver = new BatteryLevelReceiver(context, this);
        this.networkConnectedReceiver = new NetworkConnectedReceiver(context, this);
        this.sendOnlyWithWifi = new AtomicBoolean(sendOnlyWithWifi);
        this.useCompression = false;
        this.authState = authState;
        this.specificData = CacheStore.get().getSpecificData();

        dataRetention = new AtomicLong(DATA_RETENTION_DEFAULT);

        submitter = null;
        sender = null;

        statusListener = null;

        if (kafkaUrl != null) {
            doEnableSubmitter();
        } else {
            updateServerStatus(Status.DISABLED);
        }

        this.maxBytes = maxBytes;
    }

    private synchronized void updateUploadRate() {
        if (submitter != null) {
            submitter.setUploadRate(getPreferredUploadRate());
        }
    }

    private long getPreferredUploadRate() {
        if (batteryLevelReceiver.hasMinimumLevel(REDUCED_BATTERY_LEVEL)) {
            return kafkaUploadRate;
        } else {
            return kafkaUploadRate * 5;
        }
    }

    /**
     * Start submitting data to the server.
     *
     * This will not do anything if there is not already a submitter running, if it is disabled,
     * if the network is not connected or if the battery is running too low.
     */
    public synchronized void start() {
        if (isStarted()
                || authState == null
                || status == Status.DISABLED
                || !networkConnectedReceiver.hasConnection(sendOnlyWithWifi.get())
                || !batteryLevelReceiver.hasMinimumLevel(minimumBatteryLevel.get())) {
            return;
        }

        updateServerStatus(Status.CONNECTING);
        RestClient client = RestClient.global()
                .server(kafkaConfig)
                .gzipCompression(useCompression)
                .timeout(senderConnectionTimeout, TimeUnit.SECONDS)
                .build();
        this.sender = new RestSender.Builder()
                .httpClient(client)
                .schemaRetriever(schemaRetriever)
                .headers(authState.getOkHttpHeaders())
                .hasBinaryContent(hasBinaryContent)
                .build();
        this.submitter = new KafkaDataSubmitter(this, sender, kafkaRecordsSendLimit,
                getPreferredUploadRate(), authState.getUserId());
    }

    private synchronized boolean isStarted() {
        return submitter != null;
    }

    /**
     * Pause sending any data.
     * This waits for any remaining data to be sent.
     */
    public synchronized void stop() {
        if (submitter != null) {
            this.submitter.close();
            this.submitter = null;
            this.sender = null;
        }
        if (status != Status.DISABLED) {
            updateServerStatus(Status.READY);
        }
    }

    /** Do not submit any data, only cache it. If it is already disabled, this does nothing. */
    public synchronized void disableSubmitter() {
        if (status != Status.DISABLED) {
            updateServerStatus(Status.DISABLED);
            if (isStarted()) {
                stop();
            }
            networkConnectedReceiver.unregister();
            batteryLevelReceiver.unregister();
        }
    }

    /** Start submitting data. If it is already submitting data, this does nothing. */
    public synchronized void enableSubmitter() {
        if (status == Status.DISABLED) {
            doEnableSubmitter();
            start();
        }
    }

    private void doEnableSubmitter() {
        networkConnectedReceiver.register();
        batteryLevelReceiver.register();
        updateServerStatus(Status.READY);
    }

    /**
     * Sends any remaining data and closes the tables and connections.
     * @throws IOException if the tables cannot be flushed
     */
    public synchronized void close() throws IOException {
        handler = null;
        handlerThread.quitSafely();

        if (status != Status.DISABLED) {
            networkConnectedReceiver.unregister();
            batteryLevelReceiver.unregister();
        }
        if (this.submitter != null) {
            this.submitter.close();  // will also close sender
            this.submitter = null;
            this.sender = null;
        }
        for (DataCacheGroup<ObservationKey, ? extends SpecificRecord> table : tables.values()) {
            table.close();
        }
    }

    /**
     * Check the connection with the server.
     *
     * Updates will be given to any listeners registered to
     * {@link #setStatusListener(ServerStatusListener)}.
     */
    public synchronized void checkConnection() {
        if (status == Status.DISABLED) {
            return;
        }
        if (!isStarted()) {
            start();
        }
        if (isStarted()) {
            submitter.checkConnection();
        }
    }

    /**
     * Get the table of a given topic
     */
    @SuppressWarnings("unchecked")
    public <V extends SpecificRecord> DataCache<ObservationKey, V> getCache(String topic) {
        return (DataCache<ObservationKey, V>) tables.get(topic).getActiveDataCache();
    }

    @Override
    public <W extends SpecificRecord> void addMeasurement(AvroTopic<ObservationKey, W> topic, ObservationKey key, W value) {
        checkRecord(topic, key, value);
        getCache(topic.getName()).addMeasurement(key, value);
    }

    private void checkRecord(AvroTopic topic, ObservationKey key, SpecificRecord value) {
        if (!specificData.validate(topic.getKeySchema(), key)
                || !specificData.validate(topic.getValueSchema(), value)) {
            throw new IllegalArgumentException("Cannot send invalid record to topic " + topic
                    + " with {key: " + key + ", value: " + value + "}");
        }
    }

    @Override
    public void setMaximumCacheSize(int numBytes) {
        maxBytes = numBytes;
        for (DataCacheGroup<?, ?> cache : tables.values()) {
            cache.getActiveDataCache().setMaximumSize(numBytes);
        }
    }

    public List<ReadableDataCache> getCaches() {
        List<ReadableDataCache> caches = new ArrayList<>(tables.size());
        for (DataCacheGroup<?, ?> table : tables.values()) {
            caches.add(table.getActiveDataCache());
            caches.addAll(table.getDeprecatedCaches());
        }
        return caches;
    }

    public List<DataCacheGroup<?, ?>> getActiveCaches() {
        if (submitter == null) {
            return Collections.emptyList();
        } else if (networkConnectedReceiver.hasWifiOrEthernet() || !sendOverDataHighPriority.get()) {
            return new ArrayList<>(tables.values());
        } else {
            List<DataCacheGroup<?, ?>> filteredCaches = new ArrayList<>(tables.size());
            for (DataCacheGroup<?, ?> table : tables.values()) {
                if (highPriorityTopics.contains(table.getTopicName())) {
                    filteredCaches.add(table);
                }
            }
            return filteredCaches;
        }
    }

    /** Add a listener for ServerStatus updates. */
    public void setStatusListener(ServerStatusListener listener) {
        synchronized (STATUS_SYNC) {
            statusListener = listener;
        }
    }

    @Override
    public void updateServerStatus(ServerStatusListener.Status status) {
        synchronized (STATUS_SYNC) {
            if (statusListener != null) {
                statusListener.updateServerStatus(status);
            }
            this.status = status;
        }
    }

    /** Get the latest server status. */
    public ServerStatusListener.Status getStatus() {
        synchronized (STATUS_SYNC) {
            return this.status;
        }
    }

    @Override
    public void updateRecordsSent(String topicName, int numberOfRecords) {
        synchronized (STATUS_SYNC) {
            if (statusListener != null) {
                statusListener.updateRecordsSent(topicName, numberOfRecords);
            }
            // Overwrite key-value if exists. Only stores the last
            this.lastNumberOfRecordsSent.put(topicName, numberOfRecords );

            if (logger.isInfoEnabled()) {
                if (numberOfRecords < 0) {
                    String message = String.format(Locale.US, "%1$45s has FAILED uploading", topicName);
                    logger.warn(message);
                } else {
                    String message = String.format(Locale.US, "%1$45s uploaded %2$4d records",
                            topicName, numberOfRecords);
                    logger.info(message);
                }
            }
        }
    }

    public Map<String, Integer> getRecordsSent() {
        synchronized (STATUS_SYNC) {
            return this.lastNumberOfRecordsSent;
        }
    }

    public void setDatabaseCommitRate(long period) {
        for (DataCacheGroup<?, ?> table : tables.values()) {
            table.getActiveDataCache().setTimeWindow(period);
        }
    }

    public synchronized void setKafkaRecordsSendLimit(int kafkaRecordsSendLimit) {
        if (submitter != null) {
            submitter.setSendLimit(kafkaRecordsSendLimit);
        }
        this.kafkaRecordsSendLimit = kafkaRecordsSendLimit;
    }

    public synchronized void setKafkaUploadRate(long kafkaUploadRate) {
        this.kafkaUploadRate = kafkaUploadRate;
        updateUploadRate();
    }

    public synchronized void setAuthState(AppAuthState state) {
        boolean authStateWasNull = this.authState == null;
        this.authState = state;
        logger.info("Updating data handler auth state to {}", authState.getOkHttpHeaders());
        if (sender != null) {
            sender.setHeaders(authState.getOkHttpHeaders());
        }
        if (submitter != null) {
            submitter.setUserId(authState.getUserId());
        }
        if (authStateWasNull) {
            start();
        }
    }

    public synchronized void setHasBinaryContent(boolean hasBinaryContent) {
        if (this.hasBinaryContent == hasBinaryContent) {
            return;
        }
        this.hasBinaryContent = hasBinaryContent;
        if (sender != null) {
            if (hasBinaryContent) {
                sender.useLegacyEncoding(RestSender.KAFKA_REST_ACCEPT_ENCODING, RestSender.KAFKA_REST_BINARY_ENCODING, true);
            } else {
                sender.useLegacyEncoding(RestSender.KAFKA_REST_ACCEPT_ENCODING, RestSender.KAFKA_REST_AVRO_ENCODING, false);
            }
        }
    }

    public synchronized void setSenderConnectionTimeout(long senderConnectionTimeout) {
        if (sender != null) {
            sender.setConnectionTimeout(senderConnectionTimeout, TimeUnit.SECONDS);
        }
        this.senderConnectionTimeout = senderConnectionTimeout;
    }

    public synchronized void setKafkaConfig(@NonNull ServerConfig kafkaUrl) {
        if (sender != null) {
            sender.setKafkaConfig(kafkaUrl);
        }
        this.kafkaConfig = kafkaUrl;
    }

    public synchronized void setSchemaRetriever(@NonNull SchemaRetriever schemaRetriever) {
        if (sender != null) {
            sender.setSchemaRetriever(schemaRetriever);
        }
        this.schemaRetriever = schemaRetriever;
    }

    public synchronized void setCompression(boolean useCompression) {
        if (sender != null) {
            sender.setCompression(useCompression);
        }
        this.useCompression = useCompression;
    }

    public void setDataRetention(long dataRetention) {
        this.dataRetention.set(dataRetention);
    }

    public void setMinimumBatteryLevel(float minimumBatteryLevel) {
        this.minimumBatteryLevel.set(minimumBatteryLevel);
        start();
    }

    public void setSendOnlyWithWifi(boolean sendOnlyWithWifi) {
        this.sendOnlyWithWifi.set(sendOnlyWithWifi);
        // trigger network connection setting update
        this.onNetworkConnectionChanged(
                networkConnectedReceiver.isConnected(),
                networkConnectedReceiver.hasWifiOrEthernet());
    }

    @Override
    public void onNetworkConnectionChanged(boolean isConnected, boolean hasWifiOrEthernet) {
        if (isStarted()) {
            if (!isConnected || (!hasWifiOrEthernet && sendOnlyWithWifi.get())) {
                logger.info("Network was disconnected, stopping data sending");
                stop();
            }
        } else {
            // Just try to start: the start method will not do anything if the parameters
            // are not right.
            start();
        }
    }

    @Override
    public void onBatteryLevelChanged(float level, boolean isPlugged) {
        if (isStarted()) {
            if (level < minimumBatteryLevel.get() && !isPlugged) {
                logger.info("Battery level getting low, stopping data sending");
                stop();
            } else {
                updateUploadRate();
            }
        } else {
            // Just try to start: the start method will not do anything if the parameters
            // are not right.
            start();
        }
    }

    public void registerTopic(AvroTopic<ObservationKey, ? extends SpecificRecord> topic) throws IOException {
        if (tables.containsKey(topic.getName())) {
            return;
        }
        DataCacheGroup<ObservationKey, ? extends SpecificRecord> cache = CacheStore.get()
                .getOrCreateCaches(context.getApplicationContext(), topic);
        cache.getActiveDataCache().setMaximumSize(maxBytes);
        tables.put(topic.getName(), cache);
    }

    public synchronized void setTopicsHighPriority(Set<String> topicsHighPriority) {
        if (!topicsHighPriority.equals(this.highPriorityTopics)) {
            this.highPriorityTopics.clear();
            this.highPriorityTopics.addAll(topicsHighPriority);
        }
    }

    public void setSendOverDataHighPriority(boolean sendOverDataHighPriority) {
        this.sendOverDataHighPriority.set(sendOverDataHighPriority);
    }
}

File: radar-commons-android/src/main/java/org/radarbase/android/data/TableDataHandler.kt
/*
 * Copyright 2017 The Hyve
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.radarbase.android.data

import android.content.Context
import io.ktor.client.HttpClient
import io.ktor.client.engine.cio.CIO
import kotlinx.coroutines.CoroutineDispatcher
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.channels.BufferOverflow
import kotlinx.coroutines.flow.*
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import org.apache.avro.specific.SpecificRecord
import org.radarbase.android.kafka.*
import org.radarbase.android.util.BatteryStageReceiver
import org.radarbase.android.util.CoroutineTaskExecutor
import org.radarbase.android.util.NetworkConnectedReceiver
import org.radarbase.kotlin.coroutines.CacheConfig
import org.radarbase.kotlin.coroutines.launchJoin
import org.radarbase.producer.io.timeout
import org.radarbase.producer.io.unsafeSsl
import org.radarbase.producer.rest.RestKafkaSender
import org.radarbase.producer.rest.RestKafkaSender.Companion.GZIP_CONTENT_ENCODING
import org.radarbase.producer.rest.RestKafkaSender.Companion.KAFKA_REST_BINARY_ENCODING
import org.radarbase.producer.rest.RestKafkaSender.Companion.KAFKA_REST_JSON_ENCODING
import org.radarbase.producer.rest.RestKafkaSender.Companion.restKafkaSender
import org.radarbase.topic.AvroTopic
import org.radarcns.kafka.ObservationKey
import org.slf4j.LoggerFactory
import java.io.IOException
import java.util.*
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.ConcurrentMap
import kotlin.collections.HashSet
import kotlin.time.Duration.Companion.hours
import kotlin.time.Duration.Companion.seconds

/**
 * Stores data in databases and sends it to the server. If kafkaConfig is null, data will only be
 * stored to disk, not uploaded.
 */
class TableDataHandler(
    private val context: Context,
    private val cacheStore: CacheStore,
    handlerDispatcher: CoroutineDispatcher = Dispatchers.Default,
) : DataHandler<ObservationKey, SpecificRecord> {
    private val tables: ConcurrentMap<String, DataCacheGroup<*, *>> = ConcurrentHashMap()
    private var config = DataHandlerConfiguration()

    private val batteryLevelReceiver: BatteryStageReceiver
    private val networkConnectedReceiver: NetworkConnectedReceiver = NetworkConnectedReceiver(context)

    override var serverStatus: MutableStateFlow<ServerStatus> = MutableStateFlow(ServerStatus.DISCONNECTED)
    override val numberOfRecords: MutableSharedFlow<CacheSize> = MutableSharedFlow(
        replay = 1000,
        onBufferOverflow = BufferOverflow.DROP_OLDEST
    )
    override val recordsSent: MutableSharedFlow<TopicSendReceipt> = MutableSharedFlow(
        replay = 1000,
        onBufferOverflow = BufferOverflow.DROP_OLDEST,
    )

    private var submitter: KafkaDataSubmitter? = null
    private var sender: RestKafkaSender? = null

    private val handlerExecutor: CoroutineTaskExecutor = CoroutineTaskExecutor(TableDataHandler::class.simpleName!!, handlerDispatcher)

    private var cachedRecordObserverJob: Job? = null
    private var job: Job? = null
    private var networkStateCollectorJob: Job? = null
    private val observedCaches: MutableSet<String> = HashSet()

    private val isStarted: Boolean
        get() = submitter != null

    private val startMutex: Mutex = Mutex()

    init {
        job = SupervisorJob().also {
            handlerExecutor.start(it)
        }

        this.batteryLevelReceiver = BatteryStageReceiver(context, config.batteryStageLevels) { stage ->
                when (stage) {
                    BatteryStageReceiver.BatteryStage.FULL -> {
                        handler {
                            submitter {
                                uploadRateMultiplier = 1
                            }
                        }
                    }

                    BatteryStageReceiver.BatteryStage.REDUCED -> {
                        logger.info("Battery level getting low, reducing data sending")
                        handler {
                            submitter {
                                uploadRateMultiplier = reducedUploadMultiplier
                            }
                        }
                    }

                    BatteryStageReceiver.BatteryStage.EMPTY -> {
                        if (isStarted) {
                            logger.info("Battery level getting very low, stopping data sending")
                            pause()
                        }
                    }
                }
            }

        submitter?.close()
        submitter = null
        sender = null

        if (config.restConfig.kafkaConfig != null) {
            handlerExecutor.execute {
                doEnableSubmitter()
            }
        } else {
            logger.info("Submitter is disabled: no kafkaConfig provided in init")
            serverStatus.value = ServerStatus.DISABLED
        }
    }

    suspend fun monitor() {
        if (networkStateCollectorJob?.isActive == true) return
        networkStateCollectorJob = handlerExecutor.returnJobAndExecute {
            startNetworkMonitoring()
            networkConnectedReceiver.state?.collectLatest { state ->
                if (isStarted) {
                    if (!state.hasConnection(config.sendOnlyWithWifi)) {
                        logger.info("Network was disconnected, stopping data sending")
                        pause()
                    }
                } else {
                    // Just try to start: the start method will not do anything if the parameters
                    // are not right.
                    start()
                }
            }
        }
    }

    private suspend fun observerRecordsForCache(cache: ReadableDataCache) {
        if (observedCaches.contains(cache.readTopic.name)) {
            logger.debug("Records for cache are already being observed")
            return
        }
        cachedRecordObserverJob = handlerExecutor.returnJobAndExecute {
            cache.numberOfRecords.collect {
                logger.trace("{} records are cached for topic: {}", it, cache.readTopic.name)
                numberOfRecords.emit(
                    CacheSize(
                        cache.readTopic.name,
                        it
                    )
                )
            }
        }
        observedCaches.add(cache.readTopic.name)
    }

    private suspend fun startNetworkMonitoring() {
        networkConnectedReceiver.monitor()
    }

    /**
     * Start submitting data to the server.
     *
     * This will not do anything if there is not already a submitter running, if it is disabled,
     * if the network is not connected or if the battery is running too low.
     */
    suspend fun start() = handlerExecutor.execute {
        startMutex.withLock {
            if (isStarted
                || config.submitterConfig.userId == null
                || serverStatus.value === ServerStatus.DISABLED
                || !networkConnectedReceiver.latestState.hasConnection(config.sendOnlyWithWifi)
                || batteryLevelReceiver.stage == BatteryStageReceiver.BatteryStage.EMPTY
            ) {
                when {
                    isStarted -> logger.info("Submitter already started")

                    config.submitterConfig.userId == null ->
                        logger.info("Submitter has no user ID set. Not starting.")

                    serverStatus.value === ServerStatus.DISABLED ->
                        logger.info("Submitter has been disabled earlier. Not starting")

                    !networkConnectedReceiver.latestState.hasConnection(config.sendOnlyWithWifi) ->
                        logger.info("No network connection available. Not starting")

                    batteryLevelReceiver.stage == BatteryStageReceiver.BatteryStage.EMPTY ->
                        logger.info("Battery is empty. Not starting")
                }
                return@execute
            }
            logger.info("Starting data submitter")

            val kafkaConfig = config.restConfig.kafkaConfig ?: return@execute

            serverStatus.value = ServerStatus.CONNECTING

            val kafkaUrl: String = kafkaConfig.urlString
            logger.trace("KafkaSenderTrace: Initializing RestKafkaSender in TableDataHandler::start")
            try {
                val kafkaSender = restKafkaSender {
                    try {
                        val currentRestConfig = config.restConfig
                        baseUrl = kafkaUrl
                        headers.appendAll(currentRestConfig.headers)
                        httpClient {
                            timeout(currentRestConfig.connectionTimeout.seconds)
                            contentType = if (currentRestConfig.hasBinaryContent) {
                                KAFKA_REST_BINARY_ENCODING
                            } else {
                                KAFKA_REST_JSON_ENCODING
                            }
                            if (currentRestConfig.useCompression) {
                                contentEncoding = GZIP_CONTENT_ENCODING
                            }
                            if (currentRestConfig.unsafeKafka) {
                                unsafeSsl()
                            }
                        }
                        currentRestConfig.schemaRetrieverUrl?.let { schemaUrl ->
                            schemaRetriever(schemaUrl) {
                                schemaTimeout = CacheConfig(
                                    refreshDuration = 2.hours,
                                    retryDuration = 30.seconds
                                )
                                httpClient = HttpClient(CIO) {
                                    timeout(10.seconds)
                                }
                            }
                        }
                    } catch (ex: Exception) {
                        logger.trace("KafkaSenderTrace: Error when setting configs for RestKafkaSender in TableDataHandler::start")
                    }
                }.also {
                    sender = it
                }

                this.submitter = KafkaDataSubmitter(this@TableDataHandler, kafkaSender, config.submitterConfig)
            } catch (e: Exception) {
                logger.trace(
                    "KafkaSenderTrace:  Exception when tweaking RestKafkaSender in TableDataHandler::start: ",
                    e
                )
            }
        }
    }

    /**
     * Pause sending any data.
     * This waits for any remaining data to be sent.
     */
    fun pause() = handlerExecutor.execute {
        submitter?.close()
        submitter = null
        sender = null
        if (serverStatus.value != ServerStatus.DISABLED) {
            serverStatus.value = ServerStatus.READY
        }
    }

    /** Do not submit any data, only cache it. If it is already disabled, this does nothing.  */
    private fun disableSubmitter() = handlerExecutor.execute {
        if (serverStatus.value !== ServerStatus.DISABLED) {
            logger.info("Submitter is disabled")
            serverStatus.value = ServerStatus.DISABLED
            if (isStarted) {
                pause()
            }
            networkStateCollectorJob?.cancel()
            networkStateCollectorJob = null
            batteryLevelReceiver.unregister()
        }
    }

    /** Start submitting data. If it is already submitting data, this does nothing.  */
    private fun enableSubmitter() = handlerExecutor.execute {
        if (serverStatus.value === ServerStatus.DISABLED) {
            doEnableSubmitter()
            start()
        }
    }

    private suspend fun doEnableSubmitter() {
        logger.info("Submitter is enabled")
        serverStatus.value = ServerStatus.READY
        monitor()
        batteryLevelReceiver.register()
    }

    /**
     * Sends any remaining data and closes the tables and connections.
     * @throws IOException if the tables cannot be flushed
     */
    @Throws(IOException::class)
    suspend fun stop() {
        if (serverStatus.value !== ServerStatus.DISABLED) {
            networkStateCollectorJob?.cancel()
            networkStateCollectorJob = null
            batteryLevelReceiver.unregister()
        }
        tables.values.forEach{
            it.stop()
        }
        this.submitter?.close()
        this.submitter = null
        this.sender = null

        cachedRecordObserverJob?.cancel()
        handlerExecutor.stop()
    }

    /**
     * Get the table of a given topic
     */
    override fun getCache(topic: String): DataCache<*, *> {
        return tables[topic]?.activeDataCache ?: throw NullPointerException()
    }

    override val caches: List<ReadableDataCache>
        get() = buildList(tables.size) {
            tables.values.forEach {
                add(it.activeDataCache)
                addAll(it.deprecatedCaches)
            }
        }

    override val activeCaches: List<DataCacheGroup<*, *>>
        get() = if (submitter == null) {
            emptyList()
        } else if (networkConnectedReceiver.latestState.hasWifiOrEthernet || !config.sendOverDataHighPriority) {
            ArrayList(tables.values)
        } else {
            tables.values.filter { it.topicName in config.highPriorityTopics }
        }

    override suspend fun flushCaches(successCallback: () -> Unit, errorCallback: () -> Unit) {
        submitter
            ?.flush(successCallback, errorCallback)
            ?: errorCallback()
    }

    @Throws(IOException::class)
    override suspend fun <V : SpecificRecord> registerCache(
        topic: AvroTopic<ObservationKey, V>,
    ): DataCache<ObservationKey, V> {
        val cache = cacheStore
            .getOrCreateCaches(context.applicationContext, topic, config.cacheConfig)
            .also { tables[topic.name] = it }
            .activeDataCache

        observerRecordsForCache(cache)
        return cache
    }

    override fun handler(build: DataHandlerConfiguration.() -> Unit) = handlerExecutor.execute {
            val oldConfig = config

            config = config.copy().apply(build)
            if (config == oldConfig) {
                return@execute
            }

            if (config.restConfig.kafkaConfig != null
                && config.restConfig.schemaRetrieverUrl != null
                && config.submitterConfig.userId != null
            ) {
                enableSubmitter()
            } else {
                if (config.restConfig.kafkaConfig == null) {
                    logger.info("No kafka configuration given. Disabling submitter")
                }
                if (config.restConfig.schemaRetrieverUrl == null) {
                    logger.info("No schema registry configuration given. Disabling submitter")
                }
                if (config.submitterConfig.userId == null) {
                    logger.info("No user ID given. Disabling submitter")
                }
                disableSubmitter()
            }

            if (config.cacheConfig != oldConfig.cacheConfig) {
                tables.values.forEach { it.activeDataCache.config.value = config.cacheConfig }
            }

            if (config.submitterConfig != oldConfig.submitterConfig) {
                when {
                    config.submitterConfig.userId == null -> disableSubmitter()
                    oldConfig.submitterConfig.userId == null -> enableSubmitter()
                    else -> submitter?.config = config.submitterConfig
                }
            }

            if (config.restConfig != oldConfig.restConfig) {
                val newRest = config.restConfig
                newRest.kafkaConfig?.let { kafkaConfig ->

                    oldConfig.restConfig.hasBinaryContent != newRest.hasBinaryContent
                    logger.trace(
                        "KafkaSenderTrace: Tweaking  RestKafkaSender in TableDataHandler::handler: Is sender null? {}.",
                        sender == null
                    )

                    try {
                        sender = sender?.config {
                            baseUrl = kafkaConfig.urlString
                            headers.appendAll(newRest.headers)

                            httpClient {
                                timeout(newRest.connectionTimeout.seconds)
                                contentType = if (newRest.hasBinaryContent) {
                                    KAFKA_REST_BINARY_ENCODING
                                } else {
                                    KAFKA_REST_JSON_ENCODING
                                }
                                if (newRest.useCompression) {
                                    contentEncoding = GZIP_CONTENT_ENCODING
                                }
                                if (newRest.unsafeKafka) {
                                    unsafeSsl()
                                }
                            }
                            val schemaRetrieverUrl = newRest.schemaRetrieverUrl
                            if (schemaRetrieverUrl != oldConfig.restConfig.schemaRetrieverUrl) {
                                schemaRetrieverUrl?.let { srUrl ->
                                    schemaRetriever(srUrl) {
                                        schemaTimeout = CacheConfig(
                                            refreshDuration = 2.hours,
                                            retryDuration = 30.seconds
                                        )
                                        httpClient = HttpClient(CIO) {
                                            timeout(10.seconds)
                                        }
                                    }
                                }
                            }
                        }
                        logger.trace("KafkaSenderTrace: Completed Initializing RestKafkaSender in TableDataHandler::handler")
                    } catch (e: Exception) {
                        logger.trace(
                            "KafkaSenderTrace:  Exception when tweaking RestKafkaSender in TableDataHandler::handler: ",
                            e
                        )
                    }

                    sender?.resetConnection()
                }
            }

            batteryLevelReceiver.stageLevels = config.batteryStageLevels

            networkConnectedReceiver.latestState.also { state ->
                if (isStarted) {
                    if (!state.hasConnection(config.sendOnlyWithWifi)) {
                        logger.info("Network was disconnected, stopping data sending.")
                        pause()
                    }
                } else {
                    // Just try to start: the start method will not do anything if the parameters
                    // are not right.
                    start()
                }
            }
            start()
        }

    companion object {
        private val logger = LoggerFactory.getLogger(TableDataHandler::class.java)
    }

    data class CacheSize(
        val topicName: String,
        val numberOfRecords: Long,
    )
}
